crates/chain/src/canonical_iter.rs

```diff
@@ -342,3 +342,189 @@ impl<A: Clone> CanonicalReason<A> {
+                *parent_count.entry(*txid).or_insert(0) += 1;
+            }
+        }
+
+        // Find root transactions (those with no parents in the canonical set)
+        let roots: Vec<Txid> = canonical_set
+            .iter()
+            .filter(|&&txid| !has_parents.contains(&txid))
+            .copied()
+            .collect();
+
+        // Sort the initial level
+        let mut current_level = roots;
+        Self::sort_level_by_chain_position(&mut current_level, &tx_map);
+
+        Self {
+            canonical_txs: tx_map,
+            current_level,
+            next_level: Vec::new(),
+            children_map,
+            parent_count,
+            current_index: 0,
+        }
+    }
+
+    /// Sort transactions within a level by their chain position
+    /// Confirmed transactions come first (sorted by height), then unconfirmed (sorted by last_seen)
+    fn sort_level_by_chain_position(
+        level: &mut [Txid],
+        canonical_txs: &HashMap<Txid, CanonicalTx<'a, Arc<Transaction>, A>>,
+    ) {
+        level.sort_by(|&a_txid, &b_txid| {
+            let a_tx = canonical_txs.get(&a_txid).expect("txid must exist");
+            let b_tx = canonical_txs.get(&b_txid).expect("txid must exist");
+
+            use crate::ChainPosition;
+            use core::cmp::Ordering;
+
+            match (&a_tx.chain_position, &b_tx.chain_position) {
+                // Both confirmed: sort by confirmation height
+                (
+                    ChainPosition::Confirmed {
+                        anchor: a_anchor, ..
+                    },
+                    ChainPosition::Confirmed {
+                        anchor: b_anchor, ..
+                    },
+                ) => {
+                    let a_height = a_anchor.confirmation_height_upper_bound();
+                    let b_height = b_anchor.confirmation_height_upper_bound();
+                    a_height.cmp(&b_height)
+                }
+                // Confirmed comes before unconfirmed
+                (ChainPosition::Confirmed { .. }, ChainPosition::Unconfirmed { .. }) => {
+                    Ordering::Less
+                }
+                // Unconfirmed comes after confirmed
+                (ChainPosition::Unconfirmed { .. }, ChainPosition::Confirmed { .. }) => {
+                    Ordering::Greater
+                }
+                // Both unconfirmed: sort by first_seen (earlier timestamp first)
+                (
+                    ChainPosition::Unconfirmed {
+                        first_seen: a_first_seen,
+                        ..
+                    },
+                    ChainPosition::Unconfirmed {
+                        first_seen: b_first_seen,
+                        ..
+                    },
+                ) => {
+                    // Earlier timestamps come first
+                    a_first_seen.cmp(b_first_seen)
+                }
+            }
```

Comment:
Wouldn't this be easier? Also, shouldn't we be sorting by first-seen (not last-seen)? I think the `ChainPosition` Ord impl already sorts by first-seen.

```suggestion
            a_tx.cmp(&b_tx)
```

Reply:
I think it can work, but I mean it's already using the `first_seen`.